metrics/system-meters/src/main/java/io/helidon/metrics/systemmeters/VThreadSystemMetersProvider.java

```diff
@@ -0,0 +1,148 @@
+ * <p>
+ * JFR delivers events in batches. For performance the values we track are stored as longs without
+ * concern for concurrent updates which should not happen anyway.
+ */
+public class VThreadSystemMetersProvider implements MetersProvider {
+
+    // Parts of the meter names.
+    static final String METER_NAME_PREFIX = "vthreads.";
+    static final String COUNT = "count";
+    static final String SUBMIT_FAILURES = "submitFailures";
+    static final String PINNED = "pinned";
+    static final String RECENT_PINNED = "recentPinned";
+    static final String STARTS = "starts";
+
+    private static final String METER_SCOPE = Meter.Scope.BASE;
+
+    private static final System.Logger LOGGER = System.getLogger(VThreadSystemMetersProvider.class.getName());
+    private final LazyValue<DistributionSummary> recentPinnedVirtualThreads = LazyValue.create(this::findPinned);
+    private long virtualThreadSubmitFails;
+    private long pinnedVirtualThreads;
+    private long virtualThreads;
+    private long virtualThreadStarts;
+
+    /**
+     * For service loading.
+     */
+    public VThreadSystemMetersProvider() {
+    }
+
+    @Override
+    public Collection<Meter.Builder<?, ?>> meterBuilders(MetricsFactory metricsFactory) {
+
+        var rs = new RecordingStream();
+
+        List<Meter.Builder<?, ?>> meterBuilders = new ArrayList<>(List.of(
+                Gauge.builder(METER_NAME_PREFIX + SUBMIT_FAILURES, () -> virtualThreadSubmitFails)
+                        .description("Virtual thread submit failures")
+                        .scope(METER_SCOPE),
+                Gauge.builder(METER_NAME_PREFIX + PINNED, () -> pinnedVirtualThreads)
+                        .description("Number of pinned virtual threads")
+                        .scope(METER_SCOPE),
+                DistributionSummary.builder(METER_NAME_PREFIX + RECENT_PINNED)
+                        .description("Pinned virtual thread durations")
+                        .scope(METER_SCOPE)));
+
+        listenFor(rs, Map.of("jdk.VirtualThreadSubmitFailed", this::recordSubmitFail,
+                             "jdk.VirtualThreadPinned", this::recordThreadPin));
+
+        if (metricsFactory.metricsConfig().virtualThreadCountEnabled()) {
+            meterBuilders.add(Gauge.builder(METER_NAME_PREFIX + COUNT, () -> virtualThreads)
+                                      .description("Active virtual threads")
+                                      .scope(METER_SCOPE));
+            meterBuilders.add(Gauge.builder(METER_NAME_PREFIX + STARTS, () -> virtualThreadStarts)
+                                      .description("Number of virtual thread starts")
+                                      .scope(METER_SCOPE));
+
+            listenFor(rs, Map.of("jdk.VirtualThreadStart", this::recordThreadStart,
+                                 "jdk.VirtualThreadEnd", this::recordThreadEnd));
+        }
```

Comment:
Should we also track non-virtual thread start/end? Or is that already available elsewhere?

Reply:
As noted in a response to another comment, we already report as a meter the current platform thread count. We do not report the total number of platform thread starts. That value is available from the JMX `ThreadMXBean` so we could add that.